kernel/packages/scene-system/stateful.scene.system.ts

```diff
@@ -17,47 +19,69 @@ class StatefulWebWorkerScene extends Script {
   @inject('EngineAPI')
   engine!: IEngineAPI
 
+  @inject('EnvironmentAPI')
+  environmentAPI!: EnvironmentAPI
+
   @inject('ParcelIdentity')
   parcelIdentity!: ParcelIdentity
 
   @inject('SceneStateStorageController')
   sceneStateStorage!: SceneStateStorageController
 
   private devToolsAdapter!: DevToolsAdapter
-  private renderer!: RendererStatefulActor
-  private sceneState!: SceneStateDefinition
+  private rendererActor!: RendererStatefulActor
+  private builderActor!: BuilderStatefulActor
+  private sceneActor!: SceneStatefulActor
   private eventSubscriber!: EventSubscriber
 
   constructor(transport: ScriptingTransport, opt?: ILogOpts) {
     super(transport, opt)
   }
 
   async systemDidEnable(): Promise<void> {
+    const currentRealm = await this.environmentAPI.getCurrentRealm()
     this.devToolsAdapter = new DevToolsAdapter(this.devTools)
-    const { cid: sceneId } = await this.parcelIdentity.getParcel()
-    this.renderer = new RendererStatefulActor(this.engine, sceneId)
+    const { cid: sceneId, land: land } = await this.parcelIdentity.getParcel()
+    this.rendererActor = new RendererStatefulActor(this.engine, sceneId)
     this.eventSubscriber = new EventSubscriber(this.engine)
+    this.builderActor = new BuilderStatefulActor(land, currentRealm!.domain, this.sceneStateStorage)
 
     // Fetch stored scene
-    const storedState = await this.sceneStateStorage.getStoredState(sceneId)
-    this.sceneState = storedState ? deserializeSceneState(storedState) : new SceneStateDefinition()
+    const initialState = await this.builderActor.getInititalSceneState()
+
+    this.sceneActor = new SceneStatefulActor(this.engine, initialState)
 
     // Listen to the renderer and update the local scene state
-    this.renderer.forwardChangesTo(this.sceneState)
+    this.rendererActor.forwardChangesTo(this.sceneActor)
 
     // Send the initial state ot the renderer
-    this.sceneState.sendStateTo(this.renderer)
-    this.renderer.sendInitFinished()
+    this.sceneActor.sendStateTo(this.rendererActor)
+
+    this.rendererActor.sendInitFinished()
     this.log('Sent initial load')
 
+    // Listen to scene state events
+    this.ListenToEvents(sceneId)
+  }
+
+  private ListenToEvents(sceneId: string): void {
     // Listen to storage requests
     this.eventSubscriber.on('stateEvent', ({ data }) => {
       if (data.type === 'StoreSceneState') {
         this.sceneStateStorage
-          .storeState(sceneId, serializeSceneState(this.sceneState))
+          .storeState(sceneId, serializeSceneState(this.sceneActor.getState()))
           .catch((error) => this.error(`Failed to store the scene's state`, error))
       }
     })
+
+    // Listen to save scene requests
+    this.eventSubscriber.on('stateEvent', ({ data }) => {
+      if (data.type === 'SaveSceneState') {
```

Comment:
I understand that this is called when the user exists BiW mode, right? But how will it behave in the future? Should every change be updated in the builder?

Reply:
Yes right now it is called only when the user exists BiW mode, however, in the future, this will change.

The idea is that unity will handle the saving of the scene every X seconds or when important events occurs (like exit biw, enter preview mode, ...etc)

The problem right now is that unity report the changes each X seconds, not always, so we can end with a different state in the builder API, in the kernel, and in Unity so if we move the responsibility to Unity, we can be sure that kernel and builder API will have always the same state, also at important events we can force Unity to report the changes and save. 

Let me know if this wasn't clear

Reply:
Kernel will continue to contain the current definition of the scene, but Unity will have now the responsibility to save the project